domain/application/service/charm.go

```diff
@@ -411,16 +426,34 @@ func (s *Service) GetCharmLXDProfile(ctx context.Context, id corecharm.ID) (inte
 // returned.
 func (s *Service) GetCharmArchivePath(ctx context.Context, id corecharm.ID) (string, error) {
 	if err := id.Validate(); err != nil {
-		return "", fmt.Errorf("charm id: %w", err)
+		return "", internalerrors.Errorf("charm id: %w", err)
 	}
 
 	path, err := s.st.GetCharmArchivePath(ctx, id)
 	if err != nil {
-		return "", errors.Trace(err)
+		return "", internalerrors.Errorf("getting charm archive path: %w", err)
 	}
 	return path, nil
 }
 
+func (s *Service) GetCharmArchive(ctx context.Context, id corecharm.ID) (io.ReadCloser, string, error) {
```

Comment:
Comment, also include that the hash is used to verify the result by the client.

Reply:
Oops added this on the wrong PR!